ldap/servers/slapd/backend.c

```diff
@@ -204,6 +206,108 @@ slapi_be_gettype(Slapi_Backend *be)
     return r;
 }
 
+/*
+ * This function periodically checks if it exists referrals entries
+ * in the server.
+ * This is used to accelerate SRCH request as the lookup for referrals
+ * has a negative impact on SRCH throughput
+ */
+static void
+referral_check(time_t start_time __attribute__((unused)), void *arg __attribute__((unused)))
+{
+    Slapi_Backend *be;
+    Slapi_PBlock *search_pb = NULL;
+    const char *suffix;
+    Slapi_Entry **entries = NULL;
+    char **referrals = NULL;
+    char *filter;
+    char *cookie;
+    int rc;
+    int exist_referral = 0;
+    LDAPControl **server_ctrls;
+
+    /* loop over the backends to check if a it exists a "smart" referral */
+    filter = "(objectclass=referral)";
+    be = slapi_get_first_backend(&cookie);
+    while (!exist_referral && be) {
```

Comment:
The problem with this logic is that if a referral is added then removed, then the check won't run and we'll continue to assume referrals exist.

Reply:
Not sure to understand your point.
ATM the fix set the flag per server, not per backend. So if a referral exists somewhere, then it stops iterating over the others backends and set the flag.
Now I will rework the fix to make it per backend as @progier389 suggested.

Reply:
The logic is ok, but I suspect that @Firstyear was confused by the way you named the variables:
- exist_referral for the local variable
- exist_referrals for the global one
And I also found this choice a bit confusing, IMHO you should have used more different names

Reply:
Right, it was confusing. I reworked the patch with a per suffix flag. So we no longer have a global 'exist_referrals.
Note that the new patch also suffers from non protected access to the backend flag. I think it is not a big deal even if we miss that flag setting as it will be set again few sec after.